shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs

```diff
@@ -0,0 +1,273 @@
+import Shelley.Spec.Ledger.BaseTypes (StrictMaybe)
+import Shelley.Spec.Ledger.Coin (Coin (..))
+import Shelley.Spec.Ledger.Keys (Hash)
+import Shelley.Spec.Ledger.MetaData (MetaDataHash)
+import Shelley.Spec.Ledger.PParams (Update)
+import Shelley.Spec.Ledger.Serialization (encodeFoldable)
+import Shelley.Spec.Ledger.TxBody (DCert (..), Wdrl (..))
+
+-- =========================================================================================================
+-- What do we require of TxBody? That it lives on the Block Chain, so it must have these instances.
+
+type ChainData t = (Eq t, Ord t, Show t, FromCBOR t, ToCBOR t, NoThunks t, NFData t, Typeable t)
+
+--type Value era = AbstractValue.Value era
+
+-- ================================================
+-- We cannot use the TxId from Shelley.Spec.Ledger.TxBody, because TxId and TxBody
+-- are mutually recursive (through the second (phantom) type index of Hash).
+-- TxId -> TxBody -> TxIn -> TxId. If we use the
+-- TxId from Shelley.Spec.Ledger.TxBody we get unsatisfiable constraints that
+-- can only be satisified by fixing an Era, at which point (TxBody era) becomes
+-- concrete and the constraint can be solved. But we want this construction to good
+-- for all Eras. So we must make the mutual recursion explicit in this file. Luckily
+-- the extra cost is small, just the TxId and TxIn declarations.
+
+newtype TxId era = TxId {_unTxId :: Hash (Crypto era) (TxBody era)}
+  deriving (Show, Eq, Ord, Generic)
+  deriving newtype (NoThunks)
+
+instance (Era era) => ToCBOR (TxId era) where
+  toCBOR (TxId h) = encode (Rec TxId !> To h)
+
+instance (Era era) => FromCBOR (TxId era) where
+  fromCBOR = decode (RecD TxId <! From)
+
+-- ==================================================
+-- For the same reason we cannot use TxId, we cannot use TxIn from Shelley.Spec.Ledger.TxBody.
+-- Here the mutual recursion is the same cycle, just with a different starting place.
+-- TxIn -> TxId -> TxBody -> TxIn
+
+data TxIn era = TxIn {-# UNPACK #-} !(TxId era) {-# UNPACK #-} !Word64
+  deriving (Ord, Eq, Generic, Show, NoThunks)
+
+instance (Era era) => ToCBOR (TxIn era) where
+  toCBOR (TxIn add word) = encode (Rec TxIn !> To add !> To word)
+
+instance (Era era) => FromCBOR (TxIn era) where
+  fromCBOR = decode (RecD TxIn <! From <! From)
+
+-- ================================================
+-- We Avoid using TxOut from Shelley.Spec.Ledger.TxBody, because this involves convoluted
+-- use of the Compactible class, which only confuses things. Note that we don't know what
+-- (Value era) is, so we really can't say much about it. In order to choose a compactible
+-- instance for it if we have to know what (Value era) is, and we won't know that until a
+-- specific Era is chosen. But the code is supposed to abstract over every Era. Feel free
+-- to change this, but the intended use here is for (Value era) is to be
+-- (Cardano.Ledger.ShelleyMA.Value.Value era), which doesn't have a good compactible form,
```

Comment:
At the moment we do rely on `Value` being compactible in other places, so this may be necessary anyway. And the same arguments that led to a compact form for `Coin` may call for a compact form for `Value` (we can at least squash the Ada in the same way)

Reply:
By reusing TxId, TxIn and TxOut, all this goes away.